query/src/users/auth/auth_mgr.rs

```diff
@@ -42,28 +42,37 @@ pub enum Credential {
 }
 
 impl AuthMgr {
-    pub async fn create(cfg: Config, users: Arc<UserApiProvider>) -> Result<Self> {
+    pub async fn create(cfg: Config, user_mgr: Arc<UserApiProvider>) -> Result<Self> {
         Ok(AuthMgr {
-            users,
+            user_mgr,
             tenant: cfg.query.tenant_id.clone(),
             jwt: JwtAuthenticator::try_create(cfg).await?,
         })
     }
 
     pub async fn no_auth(&self) -> Result<UserInfo> {
-        self.users
+        self.user_mgr
             .get_user(&self.tenant, UserIdentity::new("root", "127.0.0.1"))
             .await
     }
 
     pub async fn auth(&self, credential: &Credential) -> Result<UserInfo> {
         match credential {
             Credential::Jwt { token: t } => {
-                let user_name = match &self.jwt {
-                    Some(j) => j.get_user(t.as_str())?,
+                let jwt = match &self.jwt {
+                    Some(j) => j.get_jwt_claims(t.as_str())?,
                     None => return Err(ErrorCode::AuthenticateFailure("jwt auth not configured.")),
                 };
-                self.users
+                let user_name = jwt.subject.unwrap();
+                if let Some(create_user) = jwt.custom.create_user {
+                    let tenant = create_user.tenant_id.unwrap_or_else(|| self.tenant.clone());
+                    let mut user_info = UserInfo::new(&user_name, "%", AuthInfo::JWT);
+                    for role in create_user.roles {
+                        user_info.grants.grant_role(role);
+                    }
+                    self.user_mgr.add_user(&tenant, user_info, true).await?;
+                }
+                self.user_mgr
```

Comment:
if add_user does return ok,  no need to get_user here?

Reply:
If the user already exists, we still need to get the user.

Reply:
why not return the UserInfo just created?

Reply:
> why not return the UserInfo just created?

User will not be created if user already exists, so we need to `get_user`.

Reply:
so, put it together,  Jwt only create a user if not exist.
the user may be altered later as normal.

lgtm